server/events.py

```diff
@@ -0,0 +1,159 @@
+            "connection_type" is "create".
+        room_code: The unique four-letters code that will represent the room.
+        username: The username of the user creating or joining the room.
+    """
+
+    connection_type: Literal["create", "join"]
+    difficulty: int = None
+    room_code: str
+    username: str
+
+    @validator("difficulty", pre=True, always=True)
+    def valid_difficulty(cls, value, values):
+        """Validates the difficulty based on the connection type."""
+        if values["connection_type"] == "create" and value is None:
+            raise ValueError("the difficulty must be specified when creating a room")
+        return value
+
+
+class DisconnectData(EventData):
+    """The data of a disconnection event.
+
+    Fields:
+        username: The username of the user disconnecting.
+    """
+
+    username: str
+
+
+class SyncData(EventData):
+    """The data of a sync event.
+
+    Fields:
+        code: The code that already exists in the room.
+        collaborators: The list of users that already collaborate in the room.
+    """
+
+    code: str
+    collaborators: list[str]
+
+
+class MoveData(EventData):
+    """The data of a move event.
+
+    Fields:
+        position: The new position of the cursor.
+    """
+
+    position: Position
+
+
+class ReplaceData(EventData):
+    """The data of a replace event.
+
+    Fields:
+        code: A list of modifications to the code.
+    """
+
+    code: list[Replacement]
+
+
+class ErrorData(EventData):
+    """The data of an error event.
+
+    Fields:
+        message: The error message.
+    """
+
+    message: str
+
+
+class EventRequest(BaseModel):
+    """A WebSocket request event.
+
+    This represent a request made from the client to the server.
+
+    Fields:
+        type: The type of the event.
+        data: The data of the event.
+        status_code (optional): The status code of the event. Only required when
+            sending an event from the server to the client.
+    """
+
+    type: EventType
+    data: EventData
+
+    @validator("data", pre=True)
+    def valid_data(cls, value: EventData | Mapping, values):
+        """Validates the data based on the event type."""
+        if isinstance(value, EventData):
+            value = value.dict()
+
+        match values["type"]:
+            case EventType.CONNECT:
+                value = ConnectData(**value)
+            case EventType.DISCONNECT:
+                value = DisconnectData(**value)
+            case EventType.SYNC:
+                value = SyncData(**value)
+            case EventType.REPLACE:
+                value = ReplaceData(**value)
+            case EventType.ERROR:
+                value = ErrorData(**value)
+        return value
+
+
+class EventResponse(EventRequest):
+    """A WebSocket event.
+
+    This represent a response sent from the server to the client.
+
+    Fields:
+        type: The type of the event.
+        data: The data of the event.
```

Comment:
Does this need to be included?

Reply:
Yeah, not really needed, I included it just for consistency.

Reply:
Fixed in 923ec688561195ee0fbb055adec4ea28d2e67473.